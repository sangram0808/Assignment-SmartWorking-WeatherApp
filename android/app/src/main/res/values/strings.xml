<resources>
    <string name="app_name">AssignmentSmartWorkingWeatherApp</string>
</resources>
