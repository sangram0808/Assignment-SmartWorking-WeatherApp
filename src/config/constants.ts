export const API_KEY = '8d0d4e48b2c39d1d7083064492e193bf';

