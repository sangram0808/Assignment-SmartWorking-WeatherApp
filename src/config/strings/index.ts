export const AppStrings = {
    APP_NAME: "Weather App",
    SEARCH: "Search",

    ERROR:"Oops! Something went wrong while fetching the weather."
}