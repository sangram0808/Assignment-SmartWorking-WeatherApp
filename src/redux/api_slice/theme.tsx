import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface ThemeState {
  theme: 'light' | 'dark';
}

const initialState: ThemeState = {
  theme: 'light',
};

const themeSlice = createSlice({
  name: 'theme',
  initialState,
  reducers: {
    toggleTheme: (state, action: PayloadAction<'light' | 'dark' | undefined>) => {
      if (action.payload) {
        state.theme = action.payload; // Explicitly set theme
      } else {
        state.theme = state.theme === 'light' ? 'dark' : 'light'; // Toggle theme if no payload
      }
    },
  },
});

export const { toggleTheme } = themeSlice.actions;
export default themeSlice.reducer;
