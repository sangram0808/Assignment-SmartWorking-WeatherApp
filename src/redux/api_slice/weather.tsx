import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { API_KEY } from '../../config/constants';

interface WeatherResponse {
  main: {
    temp: number;
    humidity: number;
  };
  weather: {
    description: string;
    icon: string;
  }[];
  name: string;
}

export const weatherApi = createApi({
  reducerPath: 'weatherApi',
  baseQuery: fetchBaseQuery({ baseUrl: 'https://api.openweathermap.org/data/2.5/' }),
  endpoints: (builder) => ({
    getWeatherByCity: builder.query<WeatherResponse, string>({
      query: (city) => `weather?q=${city}&appid=${API_KEY}&units=metric`,
    }),
  }),
});

export const { useGetWeatherByCityQuery } = weatherApi;
