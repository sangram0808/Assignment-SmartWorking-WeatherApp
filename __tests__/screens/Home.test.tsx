import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react-native';
import HomeScreen from '../../src/screens/home';
import { Provider } from 'react-redux';
import { setupApiStore } from './testUtils';
import { weatherApi } from '../../src/redux/api_slice/weather';
import { server } from './server';
import { http, HttpResponse } from 'msw';
import { AppStrings } from '../../src/config/strings';
import { act } from 'react-test-renderer';

const mockStore = setupApiStore(weatherApi);

describe('Weather Screen Tests', () => {

    const renderComponent = () =>
        render(
            <Provider store={mockStore}>
                <HomeScreen />
            </Provider>
        );

    test('Test 1 : Weather Screen Snapshot', () => {
        renderComponent();
        expect(screen.toJSON()).toMatchSnapshot();
    });

    test('Test 2 : Fetches and displays weather data', async () => {
        renderComponent();

        expect(screen.getByTestId('city-input')).toBeTruthy();
        expect(screen.getByText('Search')).toBeTruthy();
    });

    test('Test 3 : Updates input field when typing', async () => {
        renderComponent();

        const input = screen.getByTestId('city-input');
        await act(async () => {
            fireEvent.changeText(input, 'London');
        });

        expect(input.props.value).toBe('London');
    });

    test('Test 4 : Handles API error correctly', async () => {
        // Override the existing request handler to return a 500 error
        server.use(
            http.get('https://api.openweathermap.org/data/2.5/weather', () => {
                return HttpResponse.error(); // Correct way to return an error response
            })
        );

        const { getByTestId, getByText } = renderComponent();

        fireEvent.changeText(getByTestId('city-input'), 'InvalidCity');
        fireEvent.press(getByText('Search'));

        // Wait for the error message to appear
       
        await waitFor(() => {
            let ponge=getByText(AppStrings.ERROR)
            expect(ponge).toBeTruthy(); // Adjust based on actual error message in your UI
        });
    });

    test('Test 5 : RTK Query: fetches weather data and updates state', async () => {

        const store = setupApiStore(weatherApi);
        await store.dispatch(weatherApi.endpoints.getWeatherByCity.initiate('London'));
        await waitFor(() => {
            const state = store.getState().weatherApi.queries;
            expect(state).toBeDefined();
        });

        const key = weatherApi.endpoints.getWeatherByCity.select('London')(store.getState());
        expect(key.data).toEqual({
            name: 'London',
            main: { temp: 22 },
            weather: [{ description: 'Clear Sky' }],
        });
    });

    test('Test 6 : Displays weather data after API fetch', async () => {
        renderComponent();

        fireEvent.changeText(screen.getByTestId('city-input'), 'London');
        fireEvent.press(screen.getByText('Search'));

        await waitFor(() => expect(screen.getByTestId('city-name')).toBeTruthy());

        expect(screen.getByTestId('city-name').props.children).toContain('London');
        expect(screen.getByTestId('city-temp')).toBeTruthy();
        expect(screen.getByTestId('city-description')).toBeTruthy();
    });
 
});


