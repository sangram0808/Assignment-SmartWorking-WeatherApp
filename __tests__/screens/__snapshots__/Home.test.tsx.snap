// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Screen Tests Test 1 : Weather Screen Snapshot 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#F0F0F0",
      },
    ]
  }
>
  <RCTSafeAreaView
    style={
      [
        {
          "flex": 1,
          "marginTop": null,
        },
      ]
    }
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#F0F0F0",
              "borderBottomColor": "#4BB3D4",
              "borderBottomWidth": 0.2,
              "flexDirection": "row",
              "height": 60,
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <View
          style={
            {
              "flex": 1,
              "paddingHorizontal": 16,
            }
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "color": undefined,
                  "fontSize": 60,
                  "fontWeight": "bold",
                },
                {
                  "color": "#1F2020",
                },
                {
                  "fontFamily": "Montserrat-Medium",
                  "fontSize": 18,
                },
              ]
            }
          >
            Weather App
          </Text>
        </View>
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          onTintColor="#1E50A6"
          style={
            [
              {
                "height": 31,
                "width": 51,
              },
              {
                "marginRight": 15,
              },
            ]
          }
          thumbTintColor="#1E50A6"
          tintColor="#767577"
          value={false}
        />
      </View>
      <View
        style={
          {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          }
        }
      >
        <TextInput
          onChangeText={[Function]}
          placeholder="Enter city"
          placeholderTextColor="#374151"
          style={
            [
              {
                "borderColor": "#A0AEC0",
                "borderRadius": 5,
                "borderWidth": 1,
                "color": "#111111",
                "fontFamily": "DMSans-Regular",
                "fontSize": 16,
                "margin": 12,
                "marginTop": 20,
                "padding": 10,
              },
              {
                "borderColor": "#A0AEC0",
                "color": "#111111",
              },
            ]
          }
          testID="city-input"
          value=""
        />
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#1E50A6",
              "borderColor": undefined,
              "borderRadius": 5,
              "borderWidth": 0,
              "justifyContent": "center",
              "marginHorizontal": 12,
              "opacity": 1,
              "overflow": "hidden",
              "padding": 8,
              "width": "40%",
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#FFFFFF",
                  "fontFamily": "DMSans-ExtraBold",
                  "fontSize": 16,
                },
                undefined,
              ]
            }
          >
            Search
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
